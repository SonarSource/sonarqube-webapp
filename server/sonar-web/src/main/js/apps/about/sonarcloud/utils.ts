/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

export interface Project {
  key: string;
  measures: { [key: string]: string };
  name: string;
  organization?: { avatar: string; name: string };
}

// TODO Get this from an external source
const PROJECTS = [
  {
    key: 'org.hibernate.search.v6poc:hibernate-search-parent',
    name: 'Hibernate Search 6 POC Parent POM',
    measures: {
      bugs: '39',
      reliability_rating: '1',
      vulnerabilities: '5',
      security_rating: '2',
      code_smells: '1',
      sqale_rating: '1.0',
      coverage: '89.9',
      duplicated_lines_density: '1.1',
      ncloc: '336726',
      ncloc_language_distribution: 'java=175123;js=26382'
    },
    organization: {
      avatar: '',
      name: 'Hibernate'
    }
  },
  {
    key: 'com.github.sdorra:web-resources',
    name: 'Web Resources',
    measures: {
      bugs: '39',
      reliability_rating: '1',
      vulnerabilities: '5',
      security_rating: '2',
      code_smells: '1',
      sqale_rating: '1.0',
      coverage: '89.9',
      duplicated_lines_density: '1.1',
      ncloc: '336726',
      ncloc_language_distribution: 'java=175123;js=26382'
    },
    organization: {
      avatar: '',
      name: 'sdorra-github'
    }
  },
  {
    key: 'vyos:vyos-1x',
    name: 'vyos-1x',
    measures: {
      bugs: '39',
      reliability_rating: '1',
      vulnerabilities: '5',
      security_rating: '2',
      code_smells: '1',
      sqale_rating: '1.0',
      coverage: '89.9',
      duplicated_lines_density: '1.1',
      ncloc: '336726',
      ncloc_language_distribution: 'java=175123;js=26382'
    },
    organization: {
      avatar: '',
      name: 'vyos'
    }
  },
  {
    key: 'sonarlint-visualstudio',
    name: 'SonarLint for Visual Studio',
    measures: {
      bugs: '39',
      reliability_rating: '1',
      vulnerabilities: '5',
      security_rating: '2',
      code_smells: '1',
      sqale_rating: '1.0',
      coverage: '89.9',
      duplicated_lines_density: '1.1',
      ncloc: '336726',
      ncloc_language_distribution: 'java=175123;js=26382'
    },
    organization: {
      avatar: '',
      name: 'sonarsource'
    }
  },
  {
    key: 'sample-1',
    name: 'Dummy project',
    measures: {
      bugs: '39',
      reliability_rating: '1',
      vulnerabilities: '5',
      security_rating: '2',
      code_smells: '1',
      sqale_rating: '1.0',
      coverage: '89.9',
      duplicated_lines_density: '1.1',
      ncloc: '336726',
      ncloc_language_distribution: 'java=175123;js=26382'
    },
    organization: {
      avatar: '',
      name: 'sonarsource'
    }
  }
];

export function requestFeaturedProjects(): Promise<Project[]> {
  return Promise.resolve(PROJECTS);
}
