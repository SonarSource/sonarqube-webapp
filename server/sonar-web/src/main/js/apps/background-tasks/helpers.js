/*
 * SonarQube
 * Copyright (C) 2009-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const ONE_SECOND = 1000;
const ONE_MINUTE = 60 * ONE_SECOND;

function format (int, suffix) {
  return `${int}${suffix}`;
}

export function formatDuration (value) {
  if (!value) {
    return '';
  }
  if (value >= ONE_MINUTE) {
    const minutes = Math.round(value / ONE_MINUTE);
    return format(minutes, 'min');
  } else if (value >= ONE_SECOND) {
    const seconds = Math.round(value / ONE_SECOND);
    return format(seconds, 's');
  } else {
    return format(value, 'ms');
  }
}
