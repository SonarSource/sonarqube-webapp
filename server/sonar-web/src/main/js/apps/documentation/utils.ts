/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
export interface DocumentationEntry {
  children: DocumentationEntry[];
  name: string;
  order: string;
  relativeName: string;
  scope?: 'sonarcloud';
  title: string;
}

export function activeOrChildrenActive(root: string, entry: DocumentationEntry) {
  return root.indexOf(getEntryRoot(entry.relativeName)) === 0;
}

export function getEntryRoot(name: string) {
  if (name.endsWith('index')) {
    return name
      .split('/')
      .slice(0, -1)
      .join('/');
  }
  return name;
}

export function getEntryChildren(entries: DocumentationEntry[], root?: string) {
  return entries.filter(entry => {
    const parts = entry.relativeName.split('/');
    const depth = root ? root.split('/').length : 0;
    return (
      (!root || entry.relativeName.indexOf(root) === 0) &&
      ((parts.length === depth + 1 && parts[depth] !== 'index') || parts[depth + 1] === 'index')
    );
  });
}
