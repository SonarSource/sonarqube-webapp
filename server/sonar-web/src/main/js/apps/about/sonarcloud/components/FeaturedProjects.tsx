/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as classNames from 'classnames';
import CountUp from 'react-countup';
import { throttle } from 'lodash';
import { Link } from 'react-router';
import { Project, requestFeaturedProjects } from '../utils';
import ChevronLeftIcon from '../../../../components/icons-components/ChevronLeftIcon';
import ChevronRightIcon from '../../../../components/icons-components/ChevronRightcon';
import CoverageRating from '../../../../components/ui/CoverageRating';
import DeferredSpinner from '../../../../components/common/DeferredSpinner';
import DuplicationsRating from '../../../../components/ui/DuplicationsRating';
import Level from '../../../../components/ui/Level';
import OrganizationAvatar from '../../../../components/common/OrganizationAvatar';
import ProjectCardLanguagesContainer from '../../../projects/components/ProjectCardLanguagesContainer';
import Rating from '../../../../components/ui/Rating';
import { formatMeasure } from '../../../../helpers/measures';
import { getMetricName } from '../../../overview/utils';
import { getProjectUrl } from '../../../../helpers/urls';
import './FeaturedProjects.css';

interface State {
  loading: boolean;
  reversing: boolean;
  slides: Array<{
    order: number;
    project: Project;
  }>;
  sliding: boolean;
  translate: number;
  viewable: boolean;
}

export default class FeaturedProjects extends React.PureComponent<{}, State> {
  container?: HTMLElement | null;
  mounted = false;

  constructor(props: {}) {
    super(props);
    this.state = {
      loading: true,
      reversing: false,
      slides: [],
      sliding: false,
      translate: 0,
      viewable: false
    };
    this.fetchProjects();
    this.handleScroll = throttle(this.handleScroll, 10);
  }

  componentDidMount() {
    document.addEventListener('scroll', this.handleScroll, true);
  }

  componentWillUnmount() {
    document.removeEventListener('scroll', this.handleScroll, true);
  }

  handleScroll = () => {
    if (this.container) {
      const rect = this.container.getBoundingClientRect();
      const windowHeight = window.innerHeight || document.documentElement.clientHeight;
      if (rect.top <= windowHeight && rect.top + rect.height >= 0) {
        this.setState({ viewable: true });
      }
    }
  };

  fetchProjects = () => {
    requestFeaturedProjects()
      .then(projects =>
        // Move the last element at the begining to properly display the carousel animations
        this.setState({
          loading: false,
          slides: [projects.pop(), ...projects].map((project: Project, id) => {
            return {
              order: id,
              project
            };
          })
        })
      )
      .catch(() => {
        /* Fail silently */
      });
  };

  handlePrevClick = () => {
    this.setState(({ slides }) => ({
      reversing: true,
      sliding: true,
      slides: slides.map(slide => {
        slide.order = slide.order === slides.length - 1 ? 0 : slide.order + 1;
        return slide;
      })
    }));
    setTimeout(() => {
      this.setState({ sliding: false });
    }, 50);
  };

  handleNextClick = () => {
    this.setState(({ slides }) => ({
      reversing: false,
      sliding: true,
      slides: slides.map(slide => {
        slide.order = slide.order === 0 ? slides.length - 1 : slide.order - 1;
        return slide;
      })
    }));
    setTimeout(() => {
      this.setState({ sliding: false });
    }, 50);
  };

  render() {
    const { loading, reversing, sliding, viewable } = this.state;
    return (
      <div
        className="sc-featured-projects sc-big-spacer-bottom"
        ref={node => (this.container = node)}>
        {!loading && (
          <button
            className="js-prev sc-project-button"
            onClick={this.handlePrevClick}
            type="button">
            <ChevronLeftIcon className="spacer-left" size={32} />
          </button>
        )}
        <div className="sc-featured-projects-container">
          <div
            className={classNames('sc-featured-projects-inner', {
              reversing,
              ready: !sliding,
              loading
            })}>
            {loading && <DeferredSpinner />}
            {!loading &&
              this.state.slides.map(slide => (
                <ProjectCard
                  key={slide.project.key}
                  order={slide.order}
                  project={slide.project}
                  viewable={viewable}
                />
              ))}
          </div>
        </div>
        {!loading && (
          <button
            className="js-next sc-project-button"
            onClick={this.handleNextClick}
            type="button">
            <ChevronRightIcon className="spacer-left" size={32} />
          </button>
        )}
      </div>
    );
  }
}

interface ProjectCardProps {
  order: number;
  project: Project;
  viewable: boolean;
}

export function ProjectCard({ project, order, viewable }: ProjectCardProps) {
  return (
    <div className="sc-project-card-container" style={{ order }}>
      <Link className="sc-project-card" to={getProjectUrl(project.key)}>
        <div className="sc-project-card-header">
          {project.organization && (
            <>
              <OrganizationAvatar
                className="no-border big-spacer-bottom"
                organization={project.organization}
              />
              <p className="sc-project-card-limited" title={project.organization.name}>
                {project.organization.name}
              </p>
            </>
          )}
          <h5 className="sc-project-card-limited spacer-bottom" title={project.name}>
            {project.name}
          </h5>
          <Level level={project.measures['alert_status']} />
        </div>
        <ul className="sc-project-card-measures">
          <ProjectIssues
            measures={project.measures}
            metric="bugs"
            ratingMetric="reliability_rating"
            viewable={viewable}
          />
          <ProjectIssues
            measures={project.measures}
            metric="vulnerabilities"
            ratingMetric="security_rating"
            viewable={viewable}
          />
          <ProjectIssues
            measures={project.measures}
            metric="code_smells"
            ratingMetric="sqale_rating"
            viewable={viewable}
          />
          <li>
            <span>{getMetricName('coverage')}</span>
            <div>
              {viewable && (
                <CountUp
                  decimal="."
                  decimals={1}
                  delay={0}
                  duration={4}
                  end={parseFloat(project.measures['coverage'])}
                  suffix="%">
                  {(data: { countUpRef?: React.RefObject<HTMLHeadingElement> }) => (
                    <h6 className="display-inline-block big-spacer-right" ref={data.countUpRef}>
                      0
                    </h6>
                  )}
                </CountUp>
              )}
              <CoverageRating value={project.measures['coverage']} />
            </div>
          </li>
          <li>
            <span>{getMetricName('duplications')}</span>
            <div>
              {viewable && (
                <CountUp
                  decimal="."
                  decimals={1}
                  delay={0}
                  duration={4}
                  end={parseFloat(project.measures['duplicated_lines_density'])}
                  suffix="%">
                  {(data: { countUpRef?: React.RefObject<HTMLHeadingElement> }) => (
                    <h6 className="display-inline-block big-spacer-right" ref={data.countUpRef}>
                      0
                    </h6>
                  )}
                </CountUp>
              )}
              <DuplicationsRating value={Number(project.measures['duplicated_lines_density'])} />
            </div>
          </li>
        </ul>
        <div className="sc-mention text-left big-spacer-top">
          {formatMeasure(project.measures['ncloc'], 'SHORT_INT')} lines of code /{' '}
          <ProjectCardLanguagesContainer
            className="display-inline-block"
            distribution={project.measures['ncloc_language_distribution']}
          />
        </div>
      </Link>
    </div>
  );
}

interface ProjectIssues {
  measures: { [key: string]: string };
  metric: string;
  ratingMetric: string;
  viewable: boolean;
}

export function ProjectIssues({ measures, metric, ratingMetric, viewable }: ProjectIssues) {
  const value = parseFloat(formatMeasure(measures[metric], 'SHORT_INT'));
  return (
    <li>
      <span>{getMetricName(metric)}</span>
      <div>
        {viewable && (
          <CountUp delay={0} duration={4} end={value}>
            {(data: { countUpRef?: React.RefObject<HTMLHeadingElement> }) => (
              <h6 className="display-inline-block big-spacer-right" ref={data.countUpRef}>
                0
              </h6>
            )}
          </CountUp>
        )}
        <Rating value={measures[ratingMetric]} />
      </div>
    </li>
  );
}
