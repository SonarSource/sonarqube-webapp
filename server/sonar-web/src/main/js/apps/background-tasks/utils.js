/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/* @flow */
import { STATUSES, ALL_TYPES, CURRENTS } from './constants';
import { toShortNotSoISOString } from '../../helpers/dates';
/*:: import type { Task } from './types'; */

export function updateTask(tasks /*: Task[] */, newTask /*: Task */) {
  return tasks.map(task => (task.id === newTask.id ? newTask : task));
}

export function mapFiltersToParameters(filters /*: Object */ = {}) {
  const parameters = {};

  if (filters.status === STATUSES.ALL) {
    parameters.status = [
      STATUSES.PENDING,
      STATUSES.IN_PROGRESS,
      STATUSES.SUCCESS,
      STATUSES.FAILED,
      STATUSES.CANCELED
    ].join();
  } else if (filters.status === STATUSES.ALL_EXCEPT_PENDING) {
    parameters.status = [
      STATUSES.IN_PROGRESS,
      STATUSES.SUCCESS,
      STATUSES.FAILED,
      STATUSES.CANCELED
    ].join();
  } else {
    parameters.status = filters.status;
  }

  if (filters.taskType !== ALL_TYPES) {
    parameters.type = filters.taskType;
  }

  if (filters.currents !== CURRENTS.ALL) {
    parameters.onlyCurrents = true;
  }

  if (filters.minSubmittedAt) {
    parameters.minSubmittedAt = toShortNotSoISOString(filters.minSubmittedAt);
  }

  if (filters.maxExecutedAt) {
    parameters.maxExecutedAt = toShortNotSoISOString(filters.maxExecutedAt);
  }

  if (filters.query) {
    parameters.componentQuery = filters.query;
  }

  if (filters.lastPage !== 1) {
    parameters.p = filters.lastPage;
  }

  return parameters;
}

const ONE_SECOND = 1000;
const ONE_MINUTE = 60 * ONE_SECOND;
const ONE_HOUR = 60 * ONE_MINUTE;

function format(int, suffix) {
  return `${int}${suffix}`;
}

export function formatDuration(value /*: ?number */) {
  if (!value) {
    return '';
  }
  if (value < ONE_SECOND) {
    return format(value, 'ms');
  } else if (value < ONE_SECOND * 10) {
    const seconds = Math.floor(value / ONE_SECOND);
    const ms = value - seconds * ONE_SECOND;
    return seconds + '.' + format(ms, 's');
  } else if (value < ONE_MINUTE) {
    const seconds = Math.floor(value / ONE_SECOND);
    return format(seconds, 's');
  } else if (value < ONE_MINUTE * 10) {
    const minutes = Math.floor(value / ONE_MINUTE);
    const seconds = Math.floor((value - minutes * ONE_MINUTE) / ONE_SECOND);
    return format(minutes, 'min') + ' ' + format(seconds, 's');
  }
  const hours = Math.floor(value / ONE_HOUR);
  const minutes = Math.floor((value - hours * ONE_HOUR) / ONE_MINUTE);
  return format(hours, 'h') + ' ' + format(minutes, 'min');
}
