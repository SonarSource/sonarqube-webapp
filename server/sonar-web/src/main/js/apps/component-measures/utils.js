/*
 * SonarQube
 * Copyright (C) 2009-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import bubbles from './bubbles';
import { formatMeasure, formatMeasureVariation } from '../../helpers/measures';

export function getLeakValue (measure) {
  if (!measure) {
    return null;
  }

  const period = measure.periods ?
      measure.periods.find(period => period.index === 1) : null;

  return period ? period.value : null;
}

export function getSingleMeasureValue (measures) {
  if (!measures || !measures.length) {
    return null;
  }

  return measures[0].value;
}

export function getSingleLeakValue (measures) {
  if (!measures || !measures.length) {
    return null;
  }

  const measure = measures[0];

  const period = measure.periods ?
      measure.periods.find(period => period.index === 1) : null;

  return period ? period.value : null;
}

export function formatLeak (value, metric) {
  if (metric.key.indexOf('new_') === 0) {
    return formatMeasure(value, metric.type);
  } else {
    return formatMeasureVariation(value, metric.type);
  }
}

export function enhanceWithLeak (measures) {
  function enhanceSingle (measure) {
    return { ...measure, leak: getLeakValue(measure) };
  }

  if (Array.isArray(measures)) {
    return measures.map(enhanceSingle);
  } else {
    return enhanceSingle(measures);
  }
}

export function enhanceWithSingleMeasure (components) {
  return components
      .map(component => {
        return {
          ...component,
          value: getSingleMeasureValue(component.measures),
          leak: getSingleLeakValue(component.measures)
        };
      })
      .filter(component => component.value != null || component.leak != null);
}

export function hasHistory (metricKey) {
  return metricKey.indexOf('new_') !== 0;
}

export function hasBubbleChart (metricKey) {
  return !!bubbles[metricKey];
}

export function hasTreemap (metric) {
  return ['PERCENT', 'RATING', 'LEVEL'].indexOf(metric.type) !== -1;
}
