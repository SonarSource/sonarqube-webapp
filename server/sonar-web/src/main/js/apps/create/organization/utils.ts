/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { translateWithParameters } from '../../../helpers/l10n';
import { formatMeasure } from '../../../helpers/measures';

export function formatPrice(price?: number, noSign?: boolean) {
  const priceFormatted = formatMeasure(price, 'FLOAT')
    .replace(/[.|,]0$/, '')
    .replace(/([.|,]\d)$/, '$10');
  return noSign ? priceFormatted : translateWithParameters('billing.price_format', priceFormatted);
}
