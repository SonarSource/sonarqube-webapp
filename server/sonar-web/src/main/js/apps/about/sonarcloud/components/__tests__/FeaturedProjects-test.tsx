/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import FeaturedProjects, { ProjectCard, ProjectIssues } from '../FeaturedProjects';
import { requestFeaturedProjects } from '../../utils';
import { click, waitAndUpdate } from '../../../../../helpers/testUtils';

jest.mock('../../utils', () => ({
  requestFeaturedProjects: jest.fn(() =>
    Promise.resolve([
      {
        key: 'foo',
        measures: { coverage: '20', foo: '15' },
        name: 'Foo',
        organization: { name: 'Foo', avatar: '' }
      },
      {
        key: 'bar',
        measures: { bar: '20', foo: '15' },
        name: 'Bar',
        organization: { name: 'Bar', avatar: '' }
      },
      {
        key: 'baz',
        measures: { bar: '20', foo: '15' },
        name: 'Baz',
        organization: { name: 'Baz', avatar: '' }
      },
      {
        key: 'foobar',
        measures: { bar: '20', foo: '15' },
        name: 'Foobar',
        organization: { name: 'Foobar', avatar: '' }
      }
    ])
  )
}));

const PROJECT = {
  key: 'foo',
  measures: { bar: '20', foo: '15' },
  name: 'Foo',
  organization: { name: 'Foo', avatar: '' }
};

beforeEach(() => {
  (requestFeaturedProjects as jest.Mock<any>).mockClear();
});

it('should render ProjectIssues correctly', () => {
  expect(
    shallow(
      <ProjectIssues
        measures={{ bar: '20', foo: '15' }}
        metric="foo"
        ratingMetric="bar"
        viewable={false}
      />
    )
  ).toMatchSnapshot();
});

it('should render ProjectCard correctly', () => {
  expect(shallow(<ProjectCard order={1} project={PROJECT} viewable={false} />)).toMatchSnapshot();
});

it('should render correctly', async () => {
  const wrapper = shallow(<FeaturedProjects />);
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
});

it('should cycle through projects', async () => {
  const wrapper = shallow(<FeaturedProjects />);
  expect(wrapper.find('DeferredSpinner')).toHaveLength(1);

  await waitAndUpdate(wrapper);

  expect(wrapper.state().slides.map((slide: any) => slide.order)).toEqual([0, 1, 2, 3]);

  click(wrapper.find('.js-next'));
  expect(wrapper.state().slides.map((slide: any) => slide.order)).toEqual([3, 0, 1, 2]);

  click(wrapper.find('.js-next'));
  click(wrapper.find('.js-next'));
  expect(wrapper.state().slides.map((slide: any) => slide.order)).toEqual([1, 2, 3, 0]);

  click(wrapper.find('.js-prev'));
  click(wrapper.find('.js-prev'));
  expect(wrapper.state().slides.map((slide: any) => slide.order)).toEqual([3, 0, 1, 2]);
});
