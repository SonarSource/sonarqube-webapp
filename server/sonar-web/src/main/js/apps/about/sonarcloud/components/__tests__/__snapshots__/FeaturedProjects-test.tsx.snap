// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ProjectCard correctly 1`] = `
<div
  className="sc-project-card-container"
  style={
    Object {
      "order": 1,
    }
  }
>
  <Link
    className="sc-project-card"
    onlyActiveOnIndex={false}
    style={Object {}}
    to={
      Object {
        "pathname": "/dashboard",
        "query": Object {
          "branch": undefined,
          "id": "foo",
        },
      }
    }
  >
    <div
      className="sc-project-card-header"
    >
      <React.Fragment>
        <OrganizationAvatar
          className="no-border big-spacer-bottom"
          organization={
            Object {
              "avatar": "",
              "name": "Foo",
            }
          }
        />
        <p
          className="sc-project-card-limited"
          title="Foo"
        >
          Foo
        </p>
      </React.Fragment>
      <h5
        className="sc-project-card-limited spacer-bottom"
        title="Foo"
      >
        Foo
      </h5>
      <Level />
    </div>
    <ul
      className="sc-project-card-measures"
    >
      <ProjectIssues
        measures={
          Object {
            "bar": "20",
            "foo": "15",
          }
        }
        metric="bugs"
        ratingMetric="reliability_rating"
        viewable={false}
      />
      <ProjectIssues
        measures={
          Object {
            "bar": "20",
            "foo": "15",
          }
        }
        metric="vulnerabilities"
        ratingMetric="security_rating"
        viewable={false}
      />
      <ProjectIssues
        measures={
          Object {
            "bar": "20",
            "foo": "15",
          }
        }
        metric="code_smells"
        ratingMetric="sqale_rating"
        viewable={false}
      />
      <li>
        <span>
          overview.metric.coverage
        </span>
        <div>
          <CoverageRating />
        </div>
      </li>
      <li>
        <span>
          overview.metric.duplications
        </span>
        <div>
          <DuplicationsRating
            value={NaN}
          />
        </div>
      </li>
    </ul>
    <div
      className="sc-mention text-left big-spacer-top"
    >
       lines of code /
       
      <Connect(ProjectCardLanguages)
        className="display-inline-block"
      />
    </div>
  </Link>
</div>
`;

exports[`should render ProjectIssues correctly 1`] = `
<li>
  <span>
    overview.metric.foo
  </span>
  <div>
    <Rating
      value="20"
    />
  </div>
</li>
`;

exports[`should render correctly 1`] = `
<div
  className="sc-featured-projects sc-big-spacer-bottom"
>
  <button
    className="js-prev sc-project-button"
    onClick={[Function]}
    type="button"
  >
    <ChevronLeftIcon
      className="spacer-left"
      size={32}
    />
  </button>
  <div
    className="sc-featured-projects-container"
  >
    <div
      className="sc-featured-projects-inner ready"
    >
      <ProjectCard
        key="foobar"
        order={0}
        project={
          Object {
            "key": "foobar",
            "measures": Object {
              "bar": "20",
              "foo": "15",
            },
            "name": "Foobar",
            "organization": Object {
              "avatar": "",
              "name": "Foobar",
            },
          }
        }
        viewable={false}
      />
      <ProjectCard
        key="foo"
        order={1}
        project={
          Object {
            "key": "foo",
            "measures": Object {
              "coverage": "20",
              "foo": "15",
            },
            "name": "Foo",
            "organization": Object {
              "avatar": "",
              "name": "Foo",
            },
          }
        }
        viewable={false}
      />
      <ProjectCard
        key="bar"
        order={2}
        project={
          Object {
            "key": "bar",
            "measures": Object {
              "bar": "20",
              "foo": "15",
            },
            "name": "Bar",
            "organization": Object {
              "avatar": "",
              "name": "Bar",
            },
          }
        }
        viewable={false}
      />
      <ProjectCard
        key="baz"
        order={3}
        project={
          Object {
            "key": "baz",
            "measures": Object {
              "bar": "20",
              "foo": "15",
            },
            "name": "Baz",
            "organization": Object {
              "avatar": "",
              "name": "Baz",
            },
          }
        }
        viewable={false}
      />
    </div>
  </div>
  <button
    className="js-next sc-project-button"
    onClick={[Function]}
    type="button"
  >
    <ChevronRightIcon
      className="spacer-left"
      size={32}
    />
  </button>
</div>
`;
